### DISCUSSION TITLE
get item by unique field other then id

### DISCUSSION THREAD

**Discussion by terion-name** [directus/directus#10644] [Feature Requests]
The fact that api can fetch unique items only by id is rendering such things as slugs useless. Really there are a lot of cases when one needs to get or update item not by id, but by slug, sku, external ids, etc.

Any field that is non-nullable and unique should be used as an identifier in api.

For graphql api it is simplier to design:
```graphql
Query { product(id: Int, slug: String) }: products
```
Where query field is singular form of collection name (or `products_unique` if singularizing is not a way) and as arguments has all available unique fields, accepting only one to be provided (and throwing an error if several provided)

For rest api it is more complicated but as a variant: `/items/:collection/:field/:value`, e.g. `/items/products/slug/foo`

**Comment by benhaynes** [comment]
Is there a reason not to use the [filter parameter](https://docs.directus.io/reference/query/#filter) for fetching items by other fields?

```
query {
	products(filter: { slug: { _eq: "xyz" } }) {
		id
	}
}
```

**Reply by terion-name** [reply]
it returns a collection, not a single item. for api consuming in strict environments with strong typings it is important to get single item when it should be single. it is correct from api design standpoint

**Comment by n0script22** [comment]


**Reply by benhaynes** [reply]
![CleanShot 2022-12-27 at 18 14 12](https://user-images.githubusercontent.com/522079/209735492-f486c079-56a5-459f-b8e4-ae4dbdf93885.png)

**Comment by MichaelNZ85** [comment]
I think this is a good idea. You should be able to fetch pages by slugs at the very least.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by hi-reeve** [reply]
please reopen this and take this into consideration

**Reply by davepoon** [reply]
Directus is a great platform, but retrieving a piece of data is not as simple as I initially thought.
I assumed there were functions to work like this `readItems('collection', 'field', ...)` and `readItem('collection', 'field', ...)` to get items or an item by a field in the SDK, but it only works for id.

**Reply by br41nslug** [reply]
https://docs.directus.io/reference/query.html#filter

**Reply by davepoon** [reply]
Thank you for the suggestion. I have already used the method to achieve the desired outcome. But as [terion-name](https://github.com/terion-name) mentioned, it returns an array of a single object, but not a single object.

It would be helpful if the documentation provided more examples of CRUD operations guide(rest, graphql and SDK), not a API reference. Supabase JS docs are easier to follow for these common operations.
https://supabase.com/docs/reference/javascript/select?example=getting-your-data

To get more info about how to query data in Directus at the moment.
There's no guide focusing on CRUD with rest, graphql and SDK, then I navigated to https://docs.directus.io/guides/sdk.html to find out what the SDK can do, 
then go to https://docs.directus.io/guides/sdk/getting-started.html, 
and then go to https://docs.directus.io/reference/items.html see if I can find out how to query by a field.

However, there's no more example of how to query by a field, then we had to try to take a look at other pages in the API reference,
then I tried to take a look at this https://docs.directus.io/reference/items.html, https://docs.directus.io/reference/system/collections.html, https://docs.directus.io/reference/filter-rules.html and https://docs.directus.io/reference/query.html

If still not sure, then we had to google and search on Discord to find out the answer.

Here's a suggestion to help new users build an app with Directus more easily. Please don't misunderstand me :)

Anyway, thanks for building such a great platform.

**Reply by terion-name** [reply]
> this discussion has not received at least 15 votes from the community

em....
<img width="154" alt="Снимок экрана 2023-12-23 в 00 02 43" src="https://github.com/directus/directus/assets/1060205/98816c9d-69c3-40f4-a56d-a35a83987fca">

**Reply by rijkvanzanten** [reply]
@terion-name It wasn't at the required 15 on the 90 day threshold as per our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html).
